# Data Science Lab-02

<b>Name: Gurvinder Kaur Matharu<br>
PRN: 20190802077

### AIM :
Consider two data sets given i.e. Customer Behavior and House Price Prediction.
1. Find Bivariate Association between numeric variables using Covariance and Simple Correlation for the given “House Price Prediction” Data set. Represent the results of covariance and correlation into n*n matrices. Where n is the number of numeric variables.
2. Find Bivariate Association between categorical variable “Gender” and numerical variable “Salary” using Point Biserial Correlation for the given Data set i.e. “Customer Behavior”.

### THEORY:
Covariance deals with the quantification of linear association between 2 numerical variables.<br>
Correlation is a linear association between two numerical variables whose values lie between [-1, 1].With correlation we can predict the intensity as well as the direction.<br>
Point biserial correlation deals with the association between categorical and numerical variables. For point biserial correlation the categorical variables should be true dichotomous.

In [1]:
import pandas as pd
import numpy as np

In [2]:
house_price = pd.read_csv('kc_house_data.csv')
customer_behaviour = pd.read_csv('Customer_Behaviour.csv')

In [3]:
house_price.head()

id     price  sqft_living  floors  zipcode
0  7129300520  221900.0         1180     1.0    98178
1  6414100192  538000.0         2570     2.0    98125
2  5631500400  180000.0          770     1.0    98028
3  2487200875  604000.0         1960     1.0    98136
4  1954400510  510000.0         1680     1.0    98074

In [4]:
customer_behaviour.head()

User ID  Gender  Age  Salary  Purchased
0  15624510    Male   19   19000          0
1  15810944    Male   35   20000          0
2  15668575  Female   26   43000          0
3  15603246  Female   27   57000          0
4  15804002    Male   19   76000          0

##### Question 1

In [5]:
def cov(x,y):
    n = len(x)
    s1=0
    s2=0
    h=0
    mean_x = sum(x)/n
    mean_y = sum(y)/n
    for i in range(n):
        s1 = x[i] - mean_x
        s2 = y[i] - mean_y
        h += s1*s2
    covariance = h/(n-1)
    
    return covariance

In [6]:
c1 = cov(house_price['price'], house_price['sqft_living'])
c2 = cov(house_price['price'], house_price['floors'])
c3 = cov(house_price['sqft_living'], house_price['floors'])
print('Covariance between columns price and sqft_living: ', c1)
print('Covariance between columns price and floors: ', c2)
print('Covariance between columns sqft_living and floors: ', c3)

Covariance between columns price and sqft_living:  236869900.9090375
Covariance between columns price and floors:  50938.972054024925
Covariance between columns sqft_living and floors:  175.54040197858484


In [7]:
def simple_corr(x,y):
    n=len(x)
    covariance = cov(x,y)
    s1,s2 =0,0
    h1, h2 =0,0
    mean_x = sum(x)/n
    mean_y = sum(y)/n
    for i in range(n):
        s1 += (x[i] - mean_x)**2
        s2 += (y[i] - mean_y)**2
    var_x = s1/(n-1)
    var_y = s2/(n-1)
    std_x = (var_x)**0.5
    std_y = (var_y)**0.5
    corr = (covariance)/(std_x*std_y)
    return corr

In [8]:
co1 = simple_corr(house_price['price'], house_price['sqft_living'])
co2 = simple_corr(house_price['price'], house_price['floors'])
co3 = simple_corr(house_price['floors'], house_price['sqft_living'])
co4 = simple_corr(house_price['price'], house_price['price'])
co5 = simple_corr(house_price['floors'], house_price['floors'])
co6 = simple_corr(house_price['sqft_living'], house_price['sqft_living'])
print('Correlation between columns price and sqft_living: ', co1)
print('Correlation between columns price and floors: ', co2)
print('Correlation between columns floors and sqft_living: ', co3)

Correlation between columns price and sqft_living:  0.7020437212325263
Correlation between columns price and floors:  0.25678570497549585
Correlation between columns floors and sqft_living:  0.3539492902366904


In [9]:
matrix = np.array([[co4, co1, co2],
                  [co1, co6, co3],
                  [co2, co3, co5]])
labels = ['price', 'sqft_living', 'floors']
df = pd.DataFrame(matrix, columns=labels, index=labels)
print(df)

                price  sqft_living    floors
price        1.000000     0.702044  0.256786
sqft_living  0.702044     1.000000  0.353949
floors       0.256786     0.353949  1.000000


The correlation between 'price' and 'sqft_living' is 0.702044, which suggests that they are moderately positive correlated, correlation between 'price' and 'floors' is 0.256786, which suggests that these 2 columns have a weak positive correlation. Therefore we can say that if the correlation between two numerical variables is closer to zero then they are weakly correlated with each other and if the value is close to either 1 or -1 then the numerical variables are strongly correlated.

In [10]:
house_price.corr()

id     price  sqft_living    floors   zipcode
id           1.000000 -0.016797    -0.012258  0.018525 -0.008224
price       -0.016797  1.000000     0.702044  0.256786 -0.053168
sqft_living -0.012258  0.702044     1.000000  0.353949 -0.199430
floors       0.018525  0.256786     0.353949  1.000000 -0.059121
zipcode     -0.008224 -0.053168    -0.199430 -0.059121  1.000000

##### Question 2

In [11]:
x=customer_behaviour.loc[(customer_behaviour.Gender=='Male')]
x=x.drop(['User ID', 'Age', 'Purchased'], axis=1)
x['Gender']=0
print(x)

     Gender  Salary
0         0   19000
1         0   20000
4         0   76000
5         0   58000
8         0   33000
..      ...     ...
390       0   33000
391       0   23000
393       0   42000
396       0   23000
398       0   33000

[196 rows x 2 columns]


In [12]:
y=customer_behaviour.loc[(customer_behaviour.Gender=='Female')]
y=y.drop(['User ID', 'Age', 'Purchased'], axis=1)
y['Gender']=1
print(y)

     Gender  Salary
2         1   43000
3         1   57000
6         1   84000
7         1  150000
9         1   65000
..      ...     ...
392       1   45000
394       1   59000
395       1   41000
397       1   20000
399       1   36000

[204 rows x 2 columns]


In [13]:
def pbc(x,y,z):
    nx=len(x['Salary'])
    ny=len(y['Salary'])
    nz=len(z)
    s=0
    mean_x=sum(x['Salary'])/nx
    mean_y=sum(y['Salary'])/ny
    mean_z=sum(z)/nz
    for i in range(nz):
        s = s + ((z[i] - mean_z)**2)
    std_z=(s/nz-1)**0.5
    p0 = nx/nz
    p1=ny/nz
    rpb = (mean_x-mean_y)*((p0*p1)**0.5)/std_z
    print(rpb)

In [14]:
pbc(x,y,customer_behaviour['Salary'])

-0.06043468532210468


The point biserial correlation between columns 'Gender' and 'Salary' in the 'customer_behaviour' datset is -0.060434, which suggests that there is a weak and negative relation between them; since the value is close to zero.

### CONCLUSION:
Bivariate association between the numeric variables of the 'House_Price_Prediction' Dataset were found with the help of covariance and simple correlation. Bivariate association between the categorical variable 'Gender' and numerical variable 'Salary' was calculated using Point Biserial Correlation.